requirements.in

```diff
@@ -0,0 +1,6 @@
+-c constraints.in # <-- contains known limitations
+
+staticjinja==4.1.3
+Jinja2==3.1.2
+PyYAML==6.0
+libsass==0.22.0
```

Comment:
Don't keep pinned list in this file as it's an input. The pins will be in the lockfile.

Reply:
I see what you mean and took another look to clean this up. Do we not want to explicitly declare versions for direct dependencies here as in the `requirements.in` file in the `ansible-documentation` repo?

Reply:
> Do we not want to explicitly declare versions for direct dependencies here as in the `requirements.in` file in the `ansible-documentation` repo?

If a project depends on specific features of a dependency, it's usually shown by putting a lower version boundary — the version when said feature appeared. So it'd be `dep >= 1.3` if a feature we need was implemented in v1.3 of `dep`. It's like saying “we make use of `dep`'s features that are present since v1.3” — so it's a declaration of dependency of certain functionality. This is why we don't list transitive dependencies here — we don't directly use features from them (if we did, they'd need to be listed).
By not specifying a version, this has a semantic meaning of “we don't care about the version, anything would work”.
Yet, it's different for lockfiles/constraints — these are meant for reproducibility. They guarantee that the environment is consistent across envs — dev/prod/ci/contributor. They describe the environment where the project is built, but not the project itself. Furthermore, they don't declare the feature dependencies — this information is lost there. Which is why we need to explicitly separate what things are actually needed so we can manage that file.
Lockfiles will help us with the snapshot of an env where we've built the project in the past, they also record a state of the env where the build wasn't failing — so whenever we regenerate the lockfile. We could see if something starts failing, and it'd be easy to figure out what transitive deps make it break — from the diff.

Reply:
P.S. It might be wise to add code comments with hints recording which features required those low boundaries.

Reply:
@webknjaz Thanks for the very thoughtful and detailed explanation. I know it probably seems like quite a basic thing but this is genuinely helpful. And I also think it would make sense now to put a lower version boundary on staticjinja. Cheers.